lib/src/multi_sig/multi_sig_builder.ts

```diff
@@ -0,0 +1,344 @@
+     * @returns {session} session after add other nonce commitment
+     * @memberof MultiSigBuilder
+     */
+    public addPartialSignature(
+        coSignerPublicKey: Buffer,
+        coSignerSignature: Buffer,
+    ): Buffer | undefined {
+        ow(coSignerPublicKey, ow.buffer);
+        ow(coSignerSignature, ow.buffer);
+
+        if (this.hasThisSignerSignature(coSignerSignature))
+            throw new Error(
+                "This co-signer's signature has already been added",
+            );
+
+        this.session = native.multiSig.addPartialSignature(
+            this.session,
+            coSignerPublicKey,
+            coSignerSignature,
+        );
+        this.coSignerSignatures!.push(coSignerSignature);
+
+        return this.session;
+    }
+
+    /**
+     * generate final signature, noted that it can only be taken after collected all partial signatures
+     *
+     * @returns {signature}
+     * @throws should collect all partial signature before calling sign
+     * @memberof MultiSigBuilder
+     */
+    public sign(): Buffer | undefined {
+        if (!this.hasColletedAllPartialSignatures())
+            throw new Error(
+                "Session can be final signed only after all co-signer's partial signatures are added",
+            );
+
+        this.signature = native.multiSig.sign(this.session);
+
+        return this.signature;
+    }
+
+    /**
+     * verify the final external signature or self is valid or not
+     *
+     * @param signature external signature
+     * @throws self signature is not ready for verification
+     * @returns {boolean}
+     * @memberof MultiSigBuilder
+     */
+    public verify(_signature?: Buffer): boolean {
+        let signature = _signature;
+        if (!_signature) {
+            if (!this.signature)
+                throw new Error('Own signature is not ready for verification');
+            signature = this.signature;
+        }
+
+        return native.multiSig.verify(
+            signature,
+            this.message,
+            this.signerPublicKeys,
+        );
+    }
+
+    private hasThisSignerNonceCommitment(nc: Buffer): boolean {
+        return this.coSignerNonceCommitments!.indexOf(nc) > -1;
+    }
+
+    private hasThisSignerNonce(n: Buffer): boolean {
+        return this.coSignerNonces!.indexOf(n) > -1;
+    }
+
+    private hasThisSignerSignature(s: Buffer): boolean {
+        return this.coSignerSignatures!.indexOf(s) > -1;
+    }
+
+    private hasCollectedAllNonceCommitments(): boolean {
+        return (
+            this.coSignerNonceCommitments!.length ===
+            this.totalExternalCoSigners()
+        );
+    }
+
+    private hasColletedAllNonces(): boolean {
+        return this.coSignerNonces!.length === this.totalExternalCoSigners();
+    }
+
+    private hasColletedAllPartialSignatures(): boolean {
+        return (
+            this.coSignerSignatures!.length === this.totalExternalCoSigners()
+        );
+    }
```

Comment:
Since `coSignerNonces`, `coSignerSignatures` and `coSignerNonceCommitment` is initiated as `[]`, we should be able to remove the `!` after those instance variables?